File: go-patterns/behavioral/observer/subject.go
package observer

type Subject interface {
	attach(observer Observer)
	detach(observer Observer)
	setAction(action string)
	getAction() string
	public()
}

type Secretary struct {
	observers []Observer
	action    string
}

func (s *Secretary) attach(observer Observer) {
	s.observers = append(s.observers, observer)
}

func (s *Secretary) detach(observer Observer) {
	//TODO implement me
	panic("implement me")
}

func (s *Secretary) setAction(action string) {
	s.action = action
}

func (s *Secretary) getAction() string {
	return s.action
}

func (s *Secretary) public() {
	for _, v := range s.observers {
		v.update()
	}
}

File: java-patterns/src/main/java/behavioral/observer/Subject.java
package behavioral.observer;

/**
 * Usage: 通知接口 <br/>
 * Date: 2023/5/2 22:54 <br/>
 *
 * @author <a href="mailto:hanzhang2566@foxmail.com">hanzhang</a>
 */
public interface Subject {
    void attach(Observer observer);

    void detach(Observer observer);

    void publish();

    String getAction();

    void setAction(String action);
}
